core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PushFilterThroughBoolOrAggregation.java

```diff
@@ -0,0 +1,296 @@
+                    return Result.empty();
+                }
+            }
+        }
+
+        if (!tupleDomain.isAll()) {
+            boolSymbolInTupleDomain = Optional.ofNullable(tupleDomain.getDomains().get().get(boolOrSymbol));
+            if (boolSymbolInTupleDomain.isPresent() && boolSymbolInTupleDomain.get().contains(Domain.singleValue(boolSymbolInTupleDomain.get().getType(), false))) {
+                return Result.empty();
+            }
+        }
+
+        if (boolSymbolInRemaining.isEmpty() && boolSymbolInTupleDomain.isEmpty()) {
+            return Result.empty();
+        }
+
+        // Push down the aggregation mask's symbol.
+        FilterNode source = new FilterNode(
+                context.getIdAllocator().getNextId(),
+                aggregationNode.getSource(),
+                aggregation.getMask().get().toSymbolReference());
+
+        // Remove boolOr symbol from the aggregation.
+        AggregationNode newAggregationNode = AggregationNode.builderFrom(aggregationNode)
+                .setSource(source)
+                .setAggregations(ImmutableMap.of())
+                .build();
+
+        // Add boolOr projection.
+        ProjectNode newProjectNode = new ProjectNode(
+                context.getIdAllocator().getNextId(),
+                newAggregationNode,
+                Assignments.builder()
+                        .putIdentities(source.getOutputSymbols())
+                        .put(boolOrSymbol, TRUE)
+                        .build());
+
+        // Restore identity projection if it is present in the original plan.
+        PlanNode filterSource = projectNode.map(project -> project.replaceChildren(ImmutableList.of(newProjectNode))).orElse(newProjectNode);
+
+        if (!tupleDomain.isAll()) {
+            TupleDomain<Symbol> newTupleDomain = tupleDomain.filter((symbol, domain) -> !symbol.equals(boolOrSymbol));
+            Expression newPredicate = combineConjuncts(
+                    DomainTranslator.toPredicate(newTupleDomain),
+                    TRUE);
+            if (!newPredicate.equals(TRUE)) {
+                return Result.ofPlanNode(new FilterNode(filterNode.getId(), filterSource, newPredicate));
+            }
+        }
+        return Result.ofPlanNode(filterSource);
+    }
+
+    public static boolean isGroupedBoolOrWithMask(AggregationNode aggregationNode)
+    {
+        if (!isGroupedAggregation(aggregationNode)) {
+            return false;
+        }
+
+        if (aggregationNode.getAggregations().size() != 1) {
+            return false;
+        }
+
+        Aggregation aggregation = getOnlyElement(aggregationNode.getAggregations().values());
+        if (aggregation.getMask().isEmpty() || aggregation.getFilter().isPresent()) {
+            return false;
+        }
+
+        BoundSignature signature = aggregation.getResolvedFunction().getSignature();
+        return signature.getName().equals(BOOL_OR) && aggregation.getArguments().getFirst() instanceof Reference;
```

Comment:
I don't understand why we need mask AND argument. I think we can just have argument itself